IPython/core/tests/test_async_helpers.py

```diff
@@ -0,0 +1,237 @@
+
+                # It should fail with all the values
+                for val in vals:
+                    with self.subTest((test_name, val)):
+                        msg = "Syntax error not raised for %s, %s" % (test_name, val)
+                        with self.assertRaises(SyntaxError, msg=msg):
+                            iprc_err(test_case.format(val=val))
+
+        def test_in_func_no_error(self):
+            # Test that the implementation of top-level return/yield
+            # detection isn't *too* aggressive, and works inside a function
+            func_contexts = []
+
+            func_contexts.append(('func', dedent("""
+            def f():""")))
+
+            func_contexts.append(('method', dedent("""
+            class MyClass:
+                def __init__(self):
+            """)))
+
+            func_contexts.append(('async-func', dedent("""
+            async def f():""")))
+
+            func_contexts.append(('closure', dedent("""
+            def f():
+                def g():
+            """)))
+
+            def nest_case(context, case):
+                # Detect indentation
+                lines = context.strip().splitlines()
+                prefix_len = 0
+                for c in lines[-1]:
+                    if c != ' ':
+                        break
+                    prefix_len += 1
+
+                indented_case = indent(case, ' ' * (prefix_len + 4))
+                return context + '\n' + indented_case
+
+            # Gather and run the tests
+            vals = ('return', 'yield')
+
+            success_tests = self._get_top_level_cases()
+            failure_tests = self._get_ry_syntax_errors()
+
+            for context_name, context in func_contexts:
+                # These tests should now successfully run
+                for test_name, test_case in success_tests:
+                    nested_case = nest_case(context, test_case)
+
+                    for val in vals:
+                        with self.subTest((test_name, context_name, val)):
+                            iprc_err(nested_case.format(val=val))
+
+                # These tests should still raise a SyntaxError
+                for test_name, test_case in failure_tests:
+                    nested_case = nest_case(context, test_case)
+
+                    for val in vals:
+                        with self.subTest((test_name, context_name, val)):
+                            with self.assertRaises(SyntaxError):
+                                iprc_err(nested_case.format(val=val))
+
+
+        def test_execute(self):
+            iprc(
+                """
+            import asyncio
+            await asyncio.sleep(0.001)
+            """
+            )
+
+        def test_autoawait(self):
+            ip.run_cell("%autoawait False")
+            ip.run_cell("%autoawait True")
+            iprc(
+                """
+                from asyncio import sleep
+                await.sleep(0.1)
```

Comment:
This is an example of where the test isn't working right - `await.sleep` is incorrect.

That said, it turns out that:

`iprc("""
from asyncio import sleep
await sleep(0.1)
""")

is *also* a `SyntaxError`, for inscrutable reasons.

Reply:
Found the problem:

```python
In [1]: from IPython.core.async_helpers import _asyncify                                                                                    
In [2]: from textwrap import dedent
In [3]: print(_asyncify(dedent(""" 
   ...: import asyncio 
   ...: await sleep(0.001) 
   ...: """)))                                                                                                                              

async def __wrapper__():
    try:
         import asyncio
        await sleep(0.001)

    finally:
        locals()
```

Note the erroneous indentation on the `from asyncio`. I think `_asyncify` is transforming the `\n` into a space somewhere.